File: src/SecurityPlatform.Modules.Vms/SyncClock.cs
namespace SecurityPlatform.Modules.Vms;

/// <summary>
/// Relógio mestre para playback multi-câmera sincronizado (posto de operação).
///
/// <para>
/// Contrato de tempo (idêntico ao cliente <c>wwwroot/js/sync-clock.js</c>):
/// o instante absoluto UTC reproduzido por um player é
/// <c>startedAtMs + currentTime*1000</c>, onde <c>startedAt</c> é o início do
/// segmento carregado naquele player. Para sincronizar N câmeras num mesmo
/// instante mestre <c>masterAbsMs</c>, cada player recebe
/// <c>currentTime = (masterAbsMs - startedAtMs) / 1000</c> (clampeado à duração).
/// </para>
///
/// Esta classe é a fonte de verdade do cálculo — testada de forma determinística
/// para o critério de aceite §8 #7 (4 canais sync ±200 ms). Não faz I/O.
/// </summary>
public static class SyncClock
{
    /// <summary>Tolerância padrão de deriva antes de forçar re-seek (critério §8 #7).</summary>
    public const double DefaultToleranceMs = 200;

    /// <summary>
    /// Posição (segundos) que o player de um slave deve assumir para exibir o
    /// instante absoluto <paramref name="masterAbsMs"/>, dado o início do seu
    /// segmento e a duração carregada. Clampeada a [0, duração - epsilon].
    /// Retorna <c>null</c> se o instante mestre está fora do segmento do slave
    /// (antes do início ou depois do fim) — o slave não tem imagem para aquele t.
    /// </summary>
    public static double? SlaveTargetSeconds(
        double masterAbsMs, double slaveStartedAtMs, double durationSec, double epsilonSec = 0.05)
    {
        if (!IsFinite(masterAbsMs) || !IsFinite(slaveStartedAtMs) || !IsFinite(durationSec) || durationSec <= 0)
            return null;

        var offset = (masterAbsMs - slaveStartedAtMs) / 1000.0;
        if (offset < -epsilonSec) return null;                 // instante antes do segmento
        if (offset > durationSec + epsilonSec) return null;    // instante depois do segmento

        return Math.Clamp(offset, 0, Math.Max(0, durationSec - epsilonSec));
    }

    /// <summary>Instante absoluto UTC (ms) atualmente exibido por um player.</summary>
    public static double SlaveAbsMs(double slaveStartedAtMs, double currentTimeSec)
        => slaveStartedAtMs + currentTimeSec * 1000.0;

    /// <summary>
    /// Deriva (ms) entre o instante exibido por um slave e o instante mestre.
    /// Sempre não-negativa.
    /// </summary>
    public static double DriftMs(double masterAbsMs, double slaveStartedAtMs, double currentTimeSec)
        => Math.Abs(SlaveAbsMs(slaveStartedAtMs, currentTimeSec) - masterAbsMs);

    /// <summary>Precisa re-seek? (deriva acima da tolerância).</summary>
    public static bool NeedsResync(double driftMs, double toleranceMs = DefaultToleranceMs)
        => driftMs > toleranceMs;

    /// <summary>Estado de um player auxiliar para avaliação de sincronismo.</summary>
    public readonly record struct SlaveState(int CameraId, double StartedAtMs, double DurationSec, double CurrentTimeSec);

    /// <summary>Decisão de alinhamento para um slave.</summary>
    public readonly record struct Alignment(int CameraId, bool HasFrame, double DriftMs, bool Resync, double? TargetSeconds);

    /// <summary>
    /// Avalia todos os slaves contra o instante mestre e devolve, para cada um,
    /// a deriva atual, se está fora de imagem e o seek necessário para realinhar.
    /// </summary>
    public static IReadOnlyList<Alignment> AlignAll(
        double masterAbsMs, IEnumerable<SlaveState> slaves, double toleranceMs = DefaultToleranceMs)
    {
        var result = new List<Alignment>();
        foreach (var s in slaves)
        {
            var target = SlaveTargetSeconds(masterAbsMs, s.StartedAtMs, s.DurationSec);
            if (target is null)
            {
                result.Add(new Alignment(s.CameraId, HasFrame: false, DriftMs: double.PositiveInfinity, Resync: false, TargetSeconds: null));
                continue;
            }
            var drift = DriftMs(masterAbsMs, s.StartedAtMs, s.CurrentTimeSec);
            result.Add(new Alignment(s.CameraId, HasFrame: true, DriftMs: drift, Resync: NeedsResync(drift, toleranceMs), TargetSeconds: target));
        }
        return result;
    }

    /// <summary>
    /// Pior deriva entre os slaves que têm imagem no instante mestre. Retorna 0
    /// se nenhum slave tem imagem. É a métrica avaliada no critério §8 #7.
    /// </summary>
    public static double WorstDriftMs(double masterAbsMs, IEnumerable<SlaveState> slaves)
    {
        double worst = 0;
        foreach (var a in AlignAll(masterAbsMs, slaves))
            if (a.HasFrame && a.DriftMs > worst) worst = a.DriftMs;
        return worst;
    }

    private static bool IsFinite(double v) => !double.IsNaN(v) && !double.IsInfinity(v);
}

File: src/SecurityPlatform.Api/wwwroot/js/sync-clock.js
/*
 * Relógio mestre para playback multi-câmera sincronizado (posto de operação).
 *
 * Espelha o contrato de SecurityPlatform.Modules.Vms/SyncClock.cs (fonte de
 * verdade, testada de forma determinística p/ o critério §8 #7 — 4 canais ±200 ms).
 *
 * Contrato de tempo: o instante absoluto UTC exibido por um player é
 *   startedAtMs + currentTime*1000
 * onde startedAt é o início do segmento carregado. Para sincronizar N câmeras
 * num instante mestre T, cada player recebe currentTime = (T - startedAtMs)/1000.
 */
(function (global) {
  'use strict';

  var TOLERANCE_MS = 200;

  function isFiniteNum(v) { return typeof v === 'number' && isFinite(v); }

  /** Posição (s) que o slave deve assumir p/ exibir masterAbsMs; null se fora do segmento. */
  function slaveTargetSeconds(masterAbsMs, slaveStartedAtMs, durationSec, epsilonSec) {
    epsilonSec = epsilonSec == null ? 0.05 : epsilonSec;
    if (!isFiniteNum(masterAbsMs) || !isFiniteNum(slaveStartedAtMs) ||
        !isFiniteNum(durationSec) || durationSec <= 0) return null;
    var offset = (masterAbsMs - slaveStartedAtMs) / 1000;
    if (offset < -epsilonSec) return null;
    if (offset > durationSec + epsilonSec) return null;
    return Math.min(Math.max(0, offset), Math.max(0, durationSec - epsilonSec));
  }

  function slaveAbsMs(slaveStartedAtMs, currentTimeSec) {
    return slaveStartedAtMs + currentTimeSec * 1000;
  }

  function driftMs(masterAbsMs, slaveStartedAtMs, currentTimeSec) {
    return Math.abs(slaveAbsMs(slaveStartedAtMs, currentTimeSec) - masterAbsMs);
  }

  function needsResync(drift, toleranceMs) {
    return drift > (toleranceMs == null ? TOLERANCE_MS : toleranceMs);
  }

  /**
   * Alinha uma lista de players auxiliares a masterAbsMs.
   * Cada slave: { video: HTMLVideoElement, startedAtMs: number }.
   * Re-seek só quando a deriva passa da tolerância (evita travar o vídeo).
   * Retorna a pior deriva (ms) entre os que têm imagem.
   */
  function alignPlayers(masterAbsMs, slaves, toleranceMs) {
    var worst = 0;
    for (var i = 0; i < slaves.length; i++) {
      var s = slaves[i], v = s.video;
      if (!v || !isFiniteNum(v.duration) || v.duration <= 0) continue;
      var target = slaveTargetSeconds(masterAbsMs, s.startedAtMs, v.duration);
      if (target == null) continue; // sem imagem nesse instante
      var d = driftMs(masterAbsMs, s.startedAtMs, v.currentTime);
      if (d > worst) worst = d;
      if (needsResync(d, toleranceMs)) {
        try { v.currentTime = target; } catch (e) { /* */ }
      }
    }
    return worst;
  }

  global.SyncClock = {
    TOLERANCE_MS: TOLERANCE_MS,
    slaveTargetSeconds: slaveTargetSeconds,
    slaveAbsMs: slaveAbsMs,
    driftMs: driftMs,
    needsResync: needsResync,
    alignPlayers: alignPlayers
  };
})(window);
